Handle MSVC warning when calling strncpy in ClmFile.cpp

Within ClmFile.cpp, MSVC posts a warning, requesting strncpy be switched to strncpy_s since it is a 'safer' alternative. I'd like to silence this warning.

strncpy_s differs in execution from strncpy in that it does not continue writing zeroes to pad up to count. I think we discussed wanting this behavior, writing zeros to pad to the count. I'm not sure why this matters though. In this case we initialized IndexEntries ourselves so there shouldn't be any nefarious data to overwrite inside of IndexEntries[i].filename. 

Oddly though, even after I add the header string.h on MSVC, I have to call strncpy_s without prepending std:: to it. MSVC refuses to recognize std::strncpy_s.

If there is a legitimate reason to use strncpy over strncpy_s here, I can push a preprocessor directive to ignore the warning.

Explaining strncpy_s from https://en.cppreference.com/w/c/string/byte/strncpy
> Same as (1), except that the function does not continue writing zeroes into the destination array to pad up to count, it stops after writing the terminating null character (if there was no null in the source, it writes one at dest[count] and then stops). Also, the following errors are detected at runtime and call the currently installed constraint handler function: 
> * src or dest is a null pointer 
> * destsz or count is zero or greater than RSIZE_MAX 
> * count is greater or equal destsz, but destsz is less or equal strnlen_s(src, count), in other words, truncation would occur 
> * overlap would occur between the source and the destination strings

I'd also be open to a third alternative to strncpy or strncpy_s, since they both feel a little bit out of place to me here.

From what I see in AppVeyor, this only seems to be affecting the x64 build, not the 32-bit build, which is interesting.

The error produced on AppVeyor is:
```
c:\projects\op2utility\src\archives\clmfile.cpp(296): warning C4996: 'strncpy': This function or variable may be unsafe. Consider using strncpy_s instead. To disable deprecation, use _CRT_SECURE_NO_WARNINGS. See online help for details. [C:\projects\op2utility\OP2Utility.vcxproj]
  c:\program files (x86)\windows kits\10\include\10.0.14393.0\ucrt\string.h(329): note: see declaration of 'strncpy'
  HuffLZ.cpp
```

The code in question is:
```cpp
// Copy the filename into the entry
std::strncpy(indexEntries[i].filename.data(), internalNames[i].data(), sizeof(IndexEntry::filename));
```

I suppose we could change the later part to use `sizeof(indexEntries[i].filename)`. It would still be a compile time constant, and would emphasize more strongly what we are taking the size of, rather than simply the type we are taking the size of.

Thanks for the link to the function documentation.

It does seem the original function does exactly what we want, while the newer "safe" method doesn't. It also seems there are no actual security flaws in the code that we need to be concerned about. The documentation suggests the newer "safer" method protects against some instances where a bad count is given, but we have a correctly set count.

Further, I could not get `strncpy_s` to work on Linux. I tried `<string.h>`, `<cstring>`, `<string>`, and with and without the `std::` prefix. None of them worked. I remember having cross platform issues with the `_s` functions before, where they'd work on MSVC, but not on Linux. Curiously, the page you've linked suggests they are standard, though I also noticed the documentation is for the C standard, rather than C++. Maybe this is one of those areas where the two languages differ.

I would recommend against using the newer `strncpy_s` method.

I looked briefly, and didn't see any alternative methods in the standard library that do exactly what we want. I suppose the easy way out here, is to simply silence the warning. We've manually checked it, and verified it is correct.

If we silence this, perhaps it can be done with `#pragma warning(push, errorNo)`/`#pragma warning(pop)`, wrapping the offending line, so as to only silence warnings for that one line.

Another option might be to write our own copy and pad method.

As for why we want the zero padding, it's because we're saving the image of that struct to disk, so if we don't zero pad, we could be writing unitiailized memory to disk. If you examined the file in a hex editor, and there was a short filename that didn't fill the filename buffer, we'd want to see trailing `00` bytes at the end, not garbage bytes (or worse, potentially sensitive data from the user's system).

--- a/src/Archives/ClmFile.cpp
+++ b/src/Archives/ClmFile.cpp
@@ -273,13 +273,11 @@ namespace Archives
 		uint32_t offset = headerSize + internalNames.size() * sizeof(IndexEntry);
 		for (std::size_t i = 0; i < internalNames.size(); ++i)
 		{
-#pragma warning( push )
-#pragma warning( disable : 4996 ) // Disable MSVC warning 'strncpy': This function or variable may be unsafe. Consider using strncpy_s instead.
-			// As of 28Jul18, strncpy_s was not available on Linux.
-			
+			// Disable MSVC warning 'strncpy': This function or variable may be unsafe. Consider using strncpy_s instead.
+			// As of 28Jul18, strncpy_s was not available on Linux.		
 			// Copy the filename into the entry
+#pragma warning( suppress : 4996 )
 			std::strncpy(indexEntries[i].filename.data(), internalNames[i].data(), sizeof(IndexEntry::filename));
-#pragma warning ( pop )
 
 			// Set the offset of the file
 			indexEntries[i].dataOffset = offset;

--- a/src/Archives/ClmFile.cpp
+++ b/src/Archives/ClmFile.cpp
@@ -273,8 +273,13 @@ namespace Archives
 		uint32_t offset = headerSize + internalNames.size() * sizeof(IndexEntry);
 		for (std::size_t i = 0; i < internalNames.size(); ++i)
 		{
+#pragma warning( push )
+#pragma warning( disable : 4996 ) // Disable MSVC warning 'strncpy': This function or variable may be unsafe. Consider using strncpy_s instead.
+			// As of 28Jul18, strncpy_s was not available on Linux.
+			
 			// Copy the filename into the entry
 			std::strncpy(indexEntries[i].filename.data(), internalNames[i].data(), sizeof(IndexEntry::filename));
+#pragma warning ( pop )
 
 			// Set the offset of the file
 			indexEntries[i].dataOffset = offset;

